### ISSUE TITLE
Patch leading to root device not found

### ISSUE THREAD

**Issue by UeliDeSchwert** [j0hnwang/OMEN-Transcend-16-ACPI-fix#2]
Hey there!
I'm trying to install Ubuntu 22.04 on my new HP Omen O16. 
The setup worked with `noacpi`, so then I tried applying the patch. 

I downloaded the dsdt.aml from the directory F.12-patch and copied it to /boot. 
Rebooting into Grub, I set `acpi /boot/dsdt.aml` before the `linux` line. (Note: Directly referencing `acpi /dsdt.aml` did not work, as it did not find the patch. Because all other files like initrd or vmlinuz are under the /boot directory, I figured that'd be the place where dsdt.aml would be located as well. And yes, using that, the patch is indeed found).
However, during boot, the system does not find the root disk:

![IMG_20240322_110355](https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/assets/2377156/cb3efa1c-d369-467c-92c3-95c47113e09c)

The root disk is referenced by its UUID and resides on a built-in M2-SSD (usually located at `/dev/nvme0n1p2` with `/boot` residing at `/dev/nvme0n1p1`)
Without the patch (and re-adding `noapic`) the system boots as usual.

I tried with adding `rootdelay=10` to the linux-command, verified the ID using `blkid` and made sure, that the ID required was the same as written in the grub config. 

I observed, that in the initramfs-emergency-shell, I don't have the `lsblk` command, and `blkid` does not output anything (the ID-verification I mentioned before was done in the running system). 

Any idea where that issue could come from, and what could be done against it? 
It's basically the same issue as described in https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/issues/1#issuecomment-1937709261 and https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/issues/1#issuecomment-1950979153, however I could not find any solution to my issue there. 
Any help is greatly appreciated! :D 

Some information:
Product Number: 81T35EA#UUZ
Bios version: F.12
Embedded Controller: 64.37
Intel ME: 16.1.25.2128
Intel ISH: 5.4.2.4587
GOP: 17.0.1081

**Comment by ccellist** [comment]
I can confirm this same issue both in Fedora 39 and in NixOS unstable. I didn't try other distros, but both OSs fail to find the root drive. I'm attaching an image with some additional output not available in the screenshot above.

I have no clue how to debug these things since I don't even know the process for figuring out how to craft, let alone compile, a dsdt file, but I would be happy to learn.

![20240324_172926](https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/assets/3281956/62594b7b-df4b-4c21-ac11-993f05caf2c1)

**Comment by ccellist** [comment]
Quick note: a Google search revealed this gem, so I guess I'll start reading... https://dortania.github.io/Getting-Started-With-ACPI/

**Comment by j0hnwang** [comment]
Since this problem is brought up again, I would suggest reading thru this [thread](https://bugzilla.kernel.org/show_bug.cgi?id=217863).

Kernel has timing problem with certain SSD controller, it should be fixed in version later than 6.5.5. Latest Ubuntu 22.04 release is using kernel 6.5.0 still. 

My ultimate solution is changing SSD to WD SN850X, the problem is fixed once and for all.

**Comment by UeliDeSchwert** [comment]
Update from my side:

I managed to create a version of the patch that works for my machine, including NVME support. 

The DSDT is slightly different for any different model version. My model is `16-u0750nz`. The easiest way I found was to extract the DSDT myself and adjusting it similar to the patch by @j0hnwang (see diff https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/commit/503346a236b371a1c59f1a157b9024ffd56fffaa).

I booted into a live system using the `noapic` flag. Then I extracted the DSDT table using `sudo cat /sys/firmware/acpi/tables/DSDT > dsdt.dat`. This extracts the DSDT into a dsdt.dat file. Also, `noapic` or `acpi=off` does not change this table, therefore we can use that table.

Next, we need to decompile this table. For that, I installed the `acpica-tools`, which are available via apt on Ubuntu. I then decompiled the table using `iasl -d dsdt.dat`. This creates a dsdt.dsl file with readable code. We now need to remove a certain part of the code which is responsible for the devices that don't work. (See diff above). Look for the following lines, they should be around line 37'190:
```
    Scope (_SB.PC00)
    {
        Device (IC04)
        {
            Name (_HID, "HPIC0004")  // _HID: Hardware ID
            Method (_STA, 0, NotSerialized)  // _STA: Status
            {
                Return (0x0F)
            }
        }
    }
```
Remove the whole part, then save the file. Recompile it using `iasl -tc dsdt.dsl` which creates a dsdt.aml  file.
Then copy this file to `/boot` using `sudo cp dsdt.aml /boot`.

The rest follows the step of the Readme of this repository:
First, we make sure it works. We reboot to the bootloader, again edit the current selection using "e" and add the line `acpi /boot/dsdt.aml` right before the linux line. This tells the bootloader to load our custom DSDT table. Using F10 we can boot. Now everything should boot correctly without needing to set acpi=off or noapic. 

If that works, we can make the changes permanent. Edit `/etc/grub.d/40_custom` (on ubuntu, in my case) and at the bottom of the file add `acpi /boot/dsdt.aml`. 
Then, update the bootloader using `sudo update-grub`.

Now everything should work as expected.

**Comment by j0hnwang** [comment]
Great, glad to hear it's been fixed. Thanks for sharing.

**Comment by ratulKabir** [comment]
> Update from my side:
> 
> I managed to create a version of the patch that works for my machine, including NVME support.
> 
> The DSDT is slightly different for any different model version. My model is `16-u0750nz`. The easiest way I found was to extract the DSDT myself and adjusting it similar to the patch by @j0hnwang (see diff [503346a](https://github.com/j0hnwang/OMEN-Transcend-16-ACPI-fix/commit/503346a236b371a1c59f1a157b9024ffd56fffaa)).
> 
> I booted into a live system using the `noapic` flag. Then I extracted the DSDT table using `sudo cat /sys/firmware/acpi/tables/DSDT > dsdt.dat`. This extracts the DSDT into a dsdt.dat file. Also, `noapic` or `acpi=off` does not change this table, therefore we can use that table.
> 
> Next, we need to decompile this table. For that, I installed the `acpica-tools`, which are available via apt on Ubuntu. I then decompiled the table using `iasl -d dsdt.dat`. This creates a dsdt.dsl file with readable code. We now need to remove a certain part of the code which is responsible for the devices that don't work. (See diff above). Look for the following lines, they should be around line 37'190:
> 
> ```
>     Scope (_SB.PC00)
>     {
>         Device (IC04)
>         {
>             Name (_HID, "HPIC0004")  // _HID: Hardware ID
>             Method (_STA, 0, NotSerialized)  // _STA: Status
>             {
>                 Return (0x0F)
>             }
>         }
>     }
> ```
> 
> Remove the whole part, then save the file. Recompile it using `iasl -tc dsdt.dsl` which creates a dsdt.aml file. Then copy this file to `/boot` using `sudo cp dsdt.aml /boot`.
> 
> The rest follows the step of the Readme of this repository: First, we make sure it works. We reboot to the bootloader, again edit the current selection using "e" and add the line `acpi /boot/dsdt.aml` right before the linux line. This tells the bootloader to load our custom DSDT table. Using F10 we can boot. Now everything should boot correctly without needing to set acpi=off or noapic.
> 
> If that works, we can make the changes permanent. Edit `/etc/grub.d/40_custom` (on ubuntu, in my case) and at the bottom of the file add `acpi /boot/dsdt.aml`. Then, update the bootloader using `sudo update-grub`.
> 
> Now everything should work as expected.

This worked for me. Big shoutout to @UeliDeSchwert and @j0hnwang. The following was fixed only by trying the solution from the comment quoted above:

1. :white_check_mark: Kernel 6.x wasn't loading 
2. :white_check_mark: Touchpad wasn't working
3. :white_check_mark: Audio wasn't (Both built-in speaker and wired headphones)

I am using this laptop: https://www.omen.com/us/en/laptops/2023-omen-transcend-intel.html
